include/blmc_robots/real_finger.hpp

```diff
@@ -115,105 +118,187 @@ class RealFinger : public robot_interfaces::Finger {
+            // was not found.  This should normally not happen, so no specific
+            // error handling is done here.
+            rt_printf("Homing failed, index not found.  Exit.\n");
+            exit(-1);
+        }
+
+        joint_modules_.set_zero_angles(encoder_index_angles + home_offset);
+
+        rt_printf("Home position: %f, %f, %f\n", encoder_index_angles[0],
+                  encoder_index_angles[1], encoder_index_angles[2]);
     }
 
-    {
 
-      Eigen::Vector3d starting_position;
-      starting_position << 1.5, 1.5, 3.0;
-
-      double kp = 0.4;
-      double kd = 0.0025;
-      int count = 0;
-      Eigen::Vector3d last_diff(std::numeric_limits<double>::max(),
-                                std::numeric_limits<double>::max(),
-                                std::numeric_limits<double>::max());
-      while (count < 2000) {
-        Eigen::Vector3d diff =
-            starting_position - get_observation(current_time_index()).angle;
-
-        // we implement here a small pd control at the current level
-        Eigen::Vector3d desired_torque =
-            kp * diff - kd * get_observation(current_time_index()).velocity;
-
-        // Send the current to the motor
-        TimeIndex t = append_desired_action(desired_torque);
-        wait_until_time_index(t);
-        if (count % 100 == 0) {
-          Eigen::Vector3d diff_difference = last_diff - diff;
-          if (std::abs(diff_difference.norm()) < 1e-5)
-            break;
-          last_diff = diff;
+    /**
+     * @brief Move to given goal position using PD control.
+     *
+     * @param goal_pos Angular goal position for each joint.
+     * @param kp Gain K_p for the PD controller.
+     * @param kd Gain K_d for the PD controller.
+     */
+    void move_to_position(Vector goal_pos, double kp, double kd)
+    {
+        /// \todo: this relies on the safety check in the motor right now,
+        /// which is maybe not the greatest idea. Without the velocity and
+        /// torque limitation in the motor this would be very unsafe
+
+        bool reached_goal = false;
+        int count = 0;
+        Eigen::Vector3d last_diff(std::numeric_limits<double>::max(),
+                                  std::numeric_limits<double>::max(), std::numeric_limits<double>::max());
+        while (!reached_goal && count < 2000)
+        {
+            Eigen::Vector3d diff = goal_pos - get_observation(current_time_index()).angle;
+
+            // we implement here a small pd control at the current level
+            Eigen::Vector3d desired_torque = kp * diff -
+                                             kd * joint_modules_.get_measured_velocities();
+
+            // Send the current to the motor
+            TimeIndex t = append_desired_action(desired_torque);
+            wait_until_time_index(t);
+            if (count % 100 == 0)
+            {
+                Eigen::Vector3d diff_difference = last_diff - diff;
```

Comment:
this is nothing but the velocity, right? so if the velocity is zero, it does not necessarily mean that the goal has been reached. also, we should probably wait a little bit with this check, as initially it may take some time to get moving, and then this loop would exit immediately. 

in fact it may be a good idea to do here something similar to above, to take the average velocity over some window.

Reply:
Since it is only doing this once very 100 iterations, it is kind of an average velocity.  Basically it checks if the joint did not move anymore for a considerable mount of time.
This also means that we are already waiting some time (100 iterations). It actually does the check once directly in the first iteration but since `last_diff` is initialized to max double, this should always resolve to false.

But I agree that this is a bit obscure, I'll add some comments for now and will try to make it clearer when moving it to the joint module.

Reply:
Changed the implementation to check if position error is below threshold and, at the same time, velocity is close to zero.